File: types/financials.ts
/**
 * Financial Statements Types
 *
 * Type definitions for financial statement data from SEC EDGAR filings
 */

// Statement types
export type StatementType = 'income' | 'balance_sheet' | 'cash_flow' | 'ratios';

// Timeframe options
export type Timeframe = 'quarterly' | 'annual' | 'trailing_twelve_months';

// Year-over-year change data
export type YoYChange = Record<string, number | null>;

// Financial data (key-value pairs of metric names and values)
export type FinancialData = Record<string, number | string | null>;

/**
 * Company metadata for financial statements
 */
export interface FinancialsMetadata {
  company_name: string;
  cik?: string | null;
  sic?: string | null;
}

/**
 * A single financial period (quarter or year)
 */
export interface FinancialPeriod {
  fiscal_year: number;
  fiscal_quarter?: number | null;
  period_end: string;
  filed_date?: string | null;
  data: FinancialData;
  yoy_change?: YoYChange | null;
}

/**
 * Response from financial statements API
 */
export interface FinancialsResponse {
  ticker: string;
  statement_type: StatementType;
  timeframe: Timeframe;
  periods: FinancialPeriod[];
  metadata: FinancialsMetadata;
}

/**
 * Combined response with all statement types
 */
export interface AllFinancialsResponse {
  ticker: string;
  timeframe: Timeframe;
  metadata: FinancialsMetadata;
  income: FinancialPeriod[] | null;
  balance: FinancialPeriod[] | null;
  cashflow: FinancialPeriod[] | null;
}

/**
 * API wrapper response
 */
export interface FinancialsAPIResponse {
  data: FinancialsResponse;
  meta: {
    timestamp: string;
  };
}

export interface AllFinancialsAPIResponse {
  data: AllFinancialsResponse;
  meta: {
    timestamp: string;
  };
}

/**
 * Row configuration for displaying financial data in tables
 */
export interface FinancialRowConfig {
  key: string;
  label: string;
  format: 'currency' | 'number' | 'percent';
  decimals?: number;
  bold?: boolean;
  indent?: number;
  calculated?: boolean;
  tooltip?: string;
}

/**
 * Income Statement row configuration
 */
export const incomeStatementRows: FinancialRowConfig[] = [
  { key: 'revenues', label: 'Total Revenue', format: 'currency', bold: true },
  { key: 'cost_of_revenue', label: 'Cost of Revenue', format: 'currency', indent: 1 },
  { key: 'gross_profit', label: 'Gross Profit', format: 'currency', bold: true },
  { key: 'operating_expenses', label: 'Operating Expenses', format: 'currency' },
  { key: 'selling_general_and_administrative_expenses', label: 'SG&A Expenses', format: 'currency', indent: 1 },
  { key: 'research_and_development', label: 'R&D Expenses', format: 'currency', indent: 1 },
  { key: 'operating_income_loss', label: 'Operating Income', format: 'currency', bold: true },
  { key: 'interest_expense_operating', label: 'Interest Expense', format: 'currency', indent: 1 },
  { key: 'income_loss_before_taxes', label: 'Pre-tax Income', format: 'currency' },
  { key: 'income_tax_expense_benefit', label: 'Income Tax', format: 'currency', indent: 1 },
  { key: 'net_income_loss', label: 'Net Income', format: 'currency', bold: true },
  { key: 'basic_earnings_per_share', label: 'Basic EPS', format: 'number', decimals: 2 },
  { key: 'diluted_earnings_per_share', label: 'Diluted EPS', format: 'number', decimals: 2 },
  { key: 'basic_average_shares', label: 'Basic Shares Outstanding', format: 'number', decimals: 0 },
  { key: 'diluted_average_shares', label: 'Diluted Shares Outstanding', format: 'number', decimals: 0 },
];

/**
 * Balance Sheet row configuration
 */
export const balanceSheetRows: FinancialRowConfig[] = [
  // Assets
  { key: 'assets', label: 'Total Assets', format: 'currency', bold: true },
  { key: 'current_assets', label: 'Current Assets', format: 'currency', indent: 1 },
  { key: 'cash_and_cash_equivalents', label: 'Cash & Equivalents', format: 'currency', indent: 2 },
  { key: 'accounts_receivable', label: 'Accounts Receivable', format: 'currency', indent: 2 },
  { key: 'inventory', label: 'Inventory', format: 'currency', indent: 2 },
  { key: 'prepaid_expenses', label: 'Prepaid Expenses', format: 'currency', indent: 2 },
  { key: 'noncurrent_assets', label: 'Non-current Assets', format: 'currency', indent: 1 },
  { key: 'fixed_assets', label: 'Property, Plant & Equipment', format: 'currency', indent: 2 },
  { key: 'intangible_assets', label: 'Intangible Assets', format: 'currency', indent: 2 },
  { key: 'goodwill', label: 'Goodwill', format: 'currency', indent: 2 },
  // Liabilities
  { key: 'liabilities', label: 'Total Liabilities', format: 'currency', bold: true },
  { key: 'current_liabilities', label: 'Current Liabilities', format: 'currency', indent: 1 },
  { key: 'accounts_payable', label: 'Accounts Payable', format: 'currency', indent: 2 },
  { key: 'short_term_debt', label: 'Short-term Debt', format: 'currency', indent: 2 },
  { key: 'noncurrent_liabilities', label: 'Non-current Liabilities', format: 'currency', indent: 1 },
  { key: 'long_term_debt', label: 'Long-term Debt', format: 'currency', indent: 2 },
  // Equity
  { key: 'equity', label: 'Total Equity', format: 'currency', bold: true },
  { key: 'equity_attributable_to_parent', label: "Shareholders' Equity", format: 'currency', indent: 1 },
  { key: 'retained_earnings', label: 'Retained Earnings', format: 'currency', indent: 2 },
  { key: 'common_stock', label: 'Common Stock', format: 'currency', indent: 2 },
];

/**
 * Cash Flow Statement row configuration
 */
export const cashFlowRows: FinancialRowConfig[] = [
  { key: 'net_cash_flow_from_operating_activities', label: 'Operating Cash Flow', format: 'currency', bold: true },
  { key: 'depreciation_and_amortization', label: 'Depreciation & Amortization', format: 'currency', indent: 1 },
  { key: 'net_cash_flow_from_investing_activities', label: 'Investing Cash Flow', format: 'currency', bold: true },
  { key: 'capital_expenditure', label: 'Capital Expenditure', format: 'currency', indent: 1 },
  { key: 'purchase_of_investment_securities', label: 'Investment Purchases', format: 'currency', indent: 1 },
  { key: 'sale_of_investment_securities', label: 'Investment Sales', format: 'currency', indent: 1 },
  { key: 'net_cash_flow_from_financing_activities', label: 'Financing Cash Flow', format: 'currency', bold: true },
  { key: 'payment_of_dividends', label: 'Dividends Paid', format: 'currency', indent: 1 },
  { key: 'repurchase_of_common_stock', label: 'Stock Buybacks', format: 'currency', indent: 1 },
  { key: 'issuance_of_common_stock', label: 'Stock Issuance', format: 'currency', indent: 1 },
  { key: 'issuance_of_debt', label: 'Debt Issuance', format: 'currency', indent: 1 },
  { key: 'repayment_of_debt', label: 'Debt Repayment', format: 'currency', indent: 1 },
  { key: 'net_cash_flow', label: 'Net Change in Cash', format: 'currency', bold: true },
  { key: 'free_cash_flow', label: 'Free Cash Flow', format: 'currency', bold: true, calculated: true, tooltip: 'Operating Cash Flow - Capital Expenditure' },
];

/**
 * Financial Ratios row configuration
 */
export const ratioRows: FinancialRowConfig[] = [
  // Profitability
  { key: 'return_on_equity', label: 'Return on Equity (ROE)', format: 'percent', bold: true },
  { key: 'return_on_assets', label: 'Return on Assets (ROA)', format: 'percent' },
  { key: 'return_on_invested_capital', label: 'Return on Invested Capital', format: 'percent' },
  { key: 'gross_margin', label: 'Gross Margin', format: 'percent' },
  { key: 'operating_margin', label: 'Operating Margin', format: 'percent' },
  { key: 'net_profit_margin', label: 'Net Profit Margin', format: 'percent' },
  // Liquidity
  { key: 'current_ratio', label: 'Current Ratio', format: 'number', decimals: 2, bold: true },
  { key: 'quick_ratio', label: 'Quick Ratio', format: 'number', decimals: 2 },
  // Leverage
  { key: 'debt_to_equity', label: 'Debt to Equity', format: 'number', decimals: 2, bold: true },
  { key: 'debt_to_assets', label: 'Debt to Assets', format: 'percent' },
  { key: 'interest_coverage', label: 'Interest Coverage', format: 'number', decimals: 2 },
  // Efficiency
  { key: 'asset_turnover', label: 'Asset Turnover', format: 'number', decimals: 2, bold: true },
  { key: 'inventory_turnover', label: 'Inventory Turnover', format: 'number', decimals: 2 },
  { key: 'receivables_turnover', label: 'Receivables Turnover', format: 'number', decimals: 2 },
  // Valuation
  { key: 'price_to_earnings', label: 'P/E Ratio', format: 'number', decimals: 2, bold: true },
  { key: 'price_to_book', label: 'P/B Ratio', format: 'number', decimals: 2 },
  { key: 'price_to_sales', label: 'P/S Ratio', format: 'number', decimals: 2 },
  { key: 'enterprise_value_to_ebitda', label: 'EV/EBITDA', format: 'number', decimals: 2 },
];

/**
 * Get row configuration for a specific statement type
 */
export function getRowConfigForStatementType(statementType: StatementType): FinancialRowConfig[] {
  switch (statementType) {
    case 'income':
      return incomeStatementRows;
    case 'balance_sheet':
      return balanceSheetRows;
    case 'cash_flow':
      return cashFlowRows;
    case 'ratios':
      return ratioRows;
    default:
      return [];
  }
}

/**
 * Statement type display labels
 */
export const statementTypeLabels: Record<StatementType, string> = {
  income: 'Income Statement',
  balance_sheet: 'Balance Sheet',
  cash_flow: 'Cash Flow',
  ratios: 'Ratios',
};

/**
 * Timeframe display labels
 */
export const timeframeLabels: Record<Timeframe, string> = {
  quarterly: 'Quarterly',
  annual: 'Annual',
  trailing_twelve_months: 'TTM',
};

/**
 * Format a period label (e.g., "Q4 2024" or "FY 2024")
 */
export function formatPeriodLabel(period: FinancialPeriod, timeframe: Timeframe): string {
  if (timeframe === 'quarterly' && period.fiscal_quarter) {
    return `Q${period.fiscal_quarter} ${period.fiscal_year}`;
  }
  if (timeframe === 'annual') {
    return `FY ${period.fiscal_year}`;
  }
  if (timeframe === 'trailing_twelve_months') {
    return `TTM ${period.fiscal_year}`;
  }
  return period.period_end;
}

File: backend/models/financials.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"
)

// StatementType represents the type of financial statement
type StatementType string

const (
	StatementTypeIncome       StatementType = "income"
	StatementTypeBalanceSheet StatementType = "balance_sheet"
	StatementTypeCashFlow     StatementType = "cash_flow"
	StatementTypeRatios       StatementType = "ratios"
)

// Timeframe represents the reporting period type
type Timeframe string

const (
	TimeframeQuarterly Timeframe = "quarterly"
	TimeframeAnnual    Timeframe = "annual"
	TimeframeTTM       Timeframe = "trailing_twelve_months"
)

// FinancialData represents the JSONB data stored in the financial_statements table
type FinancialData map[string]interface{}

// Value implements the driver.Valuer interface for FinancialData
func (fd FinancialData) Value() (driver.Value, error) {
	if fd == nil {
		return nil, nil
	}
	return json.Marshal(fd)
}

// Scan implements the sql.Scanner interface for FinancialData
func (fd *FinancialData) Scan(value interface{}) error {
	if value == nil {
		*fd = nil
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("type assertion to []byte failed")
	}

	return json.Unmarshal(bytes, fd)
}

// FinancialStatement represents a financial statement record in the database
type FinancialStatement struct {
	ID               int           `json:"id" db:"id"`
	TickerID         int           `json:"ticker_id" db:"ticker_id"`
	CIK              *string       `json:"cik,omitempty" db:"cik"`
	StatementType    StatementType `json:"statement_type" db:"statement_type"`
	Timeframe        Timeframe     `json:"timeframe" db:"timeframe"`
	FiscalYear       int           `json:"fiscal_year" db:"fiscal_year"`
	FiscalQuarter    *int          `json:"fiscal_quarter,omitempty" db:"fiscal_quarter"`
	PeriodStart      *time.Time    `json:"period_start,omitempty" db:"period_start"`
	PeriodEnd        time.Time     `json:"period_end" db:"period_end"`
	FiledDate        *time.Time    `json:"filed_date,omitempty" db:"filed_date"`
	SourceFilingURL  *string       `json:"source_filing_url,omitempty" db:"source_filing_url"`
	SourceFilingType *string       `json:"source_filing_type,omitempty" db:"source_filing_type"`
	Data             FinancialData `json:"data" db:"data"`
	CreatedAt        time.Time     `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time     `json:"updated_at" db:"updated_at"`
}

// FinancialPeriod represents a single period's financial data for API response
type FinancialPeriod struct {
	FiscalYear    int                    `json:"fiscal_year"`
	FiscalQuarter *int                   `json:"fiscal_quarter,omitempty"`
	PeriodEnd     string                 `json:"period_end"`
	FiledDate     *string                `json:"filed_date,omitempty"`
	Data          map[string]interface{} `json:"data"`
	YoYChange     map[string]*float64    `json:"yoy_change,omitempty"`
}

// FinancialsMetadata contains company metadata for financial statements
type FinancialsMetadata struct {
	CompanyName string  `json:"company_name"`
	CIK         *string `json:"cik,omitempty"`
	SIC         *string `json:"sic,omitempty"`
}

// FinancialsResponse represents the API response for financial statements
type FinancialsResponse struct {
	Ticker        string             `json:"ticker"`
	StatementType StatementType      `json:"statement_type"`
	Timeframe     Timeframe          `json:"timeframe"`
	Periods       []FinancialPeriod  `json:"periods"`
	Metadata      FinancialsMetadata `json:"metadata"`
}

// FinancialsParams represents query parameters for financial statements
type FinancialsParams struct {
	Ticker     string    `json:"ticker"`
	Timeframe  Timeframe `json:"timeframe"`
	Limit      int       `json:"limit"`
	FiscalYear *int      `json:"fiscal_year,omitempty"`
	Sort       string    `json:"sort"` // "asc" or "desc"
}

// PolygonFinancialsResponse represents the Polygon.io financials API response
type PolygonFinancialsResponse struct {
	Status    string                  `json:"status"`
	RequestID string                  `json:"request_id"`
	Count     int                     `json:"count"`
	Results   []PolygonFinancialsData `json:"results"`
	NextURL   *string                 `json:"next_url,omitempty"`
}

// PolygonFinancialsData represents a single financial statement from Polygon.io
type PolygonFinancialsData struct {
	Ticker           string                 `json:"ticker"`
	CIK              string                 `json:"cik"`
	CompanyName      string                 `json:"company_name"`
	StartDate        string                 `json:"start_date"`
	EndDate          string                 `json:"end_date"`
	FilingDate       string                 `json:"filing_date"`
	FiscalPeriod     string                 `json:"fiscal_period"` // "Q1", "Q2", "Q3", "Q4", "FY", "TTM"
	FiscalYear       string                 `json:"fiscal_year"`
	SourceFilingURL  string                 `json:"source_filing_url"`
	SourceFilingType string                 `json:"source_filing_type"`
	Financials       PolygonFinancialsItems `json:"financials"`
}

// PolygonFinancialsItems contains the financial statement data organized by statement type
type PolygonFinancialsItems struct {
	IncomeStatement     map[string]PolygonFinancialValue `json:"income_statement,omitempty"`
	BalanceSheet        map[string]PolygonFinancialValue `json:"balance_sheet,omitempty"`
	CashFlowStatement   map[string]PolygonFinancialValue `json:"cash_flow_statement,omitempty"`
	ComprehensiveIncome map[string]PolygonFinancialValue `json:"comprehensive_income,omitempty"`
}

// PolygonFinancialValue represents a single financial value from Polygon.io
type PolygonFinancialValue struct {
	Value float64 `json:"value"`
	Unit  string  `json:"unit"`
	Label string  `json:"label"`
	Order int     `json:"order"`
}

// PolygonRatiosResponse represents the Polygon.io ratios API response
type PolygonRatiosResponse struct {
	Status    string              `json:"status"`
	RequestID string              `json:"request_id"`
	Count     int                 `json:"count"`
	Results   []PolygonRatiosData `json:"results"`
	NextURL   *string             `json:"next_url,omitempty"`
}

// PolygonRatiosData represents financial ratios from Polygon.io
type PolygonRatiosData struct {
	Ticker           string                 `json:"ticker"`
	CIK              string                 `json:"cik"`
	CompanyName      string                 `json:"company_name"`
	StartDate        string                 `json:"start_date"`
	EndDate          string                 `json:"end_date"`
	FilingDate       string                 `json:"filing_date"`
	FiscalPeriod     string                 `json:"fiscal_period"`
	FiscalYear       string                 `json:"fiscal_year"`
	SourceFilingURL  string                 `json:"source_filing_url"`
	SourceFilingType string                 `json:"source_filing_type"`
	Ratios           map[string]interface{} `json:"ratios"`
}

// Common income statement field mappings
var IncomeStatementFields = map[string]string{
	"revenues":           "Total Revenue",
	"cost_of_revenue":    "Cost of Revenue",
	"gross_profit":       "Gross Profit",
	"operating_expenses": "Operating Expenses",
	"selling_general_and_administrative_expenses": "SG&A Expenses",
	"research_and_development":                    "R&D Expenses",
	"operating_income_loss":                       "Operating Income",
	"interest_expense_operating":                  "Interest Expense",
	"income_loss_before_taxes":                    "Pre-tax Income",
	"income_tax_expense_benefit":                  "Income Tax",
	"net_income_loss":                             "Net Income",
	"basic_earnings_per_share":                    "Basic EPS",
	"diluted_earnings_per_share":                  "Diluted EPS",
	"basic_average_shares":                        "Basic Shares Outstanding",
	"diluted_average_shares":                      "Diluted Shares Outstanding",
}

// Common balance sheet field mappings
var BalanceSheetFields = map[string]string{
	"assets":                        "Total Assets",
	"current_assets":                "Current Assets",
	"cash_and_cash_equivalents":     "Cash & Equivalents",
	"accounts_receivable":           "Accounts Receivable",
	"inventory":                     "Inventory",
	"prepaid_expenses":              "Prepaid Expenses",
	"noncurrent_assets":             "Non-current Assets",
	"fixed_assets":                  "Property, Plant & Equipment",
	"intangible_assets":             "Intangible Assets",
	"goodwill":                      "Goodwill",
	"liabilities":                   "Total Liabilities",
	"current_liabilities":           "Current Liabilities",
	"accounts_payable":              "Accounts Payable",
	"short_term_debt":               "Short-term Debt",
	"noncurrent_liabilities":        "Non-current Liabilities",
	"long_term_debt":                "Long-term Debt",
	"equity":                        "Total Equity",
	"equity_attributable_to_parent": "Shareholders' Equity",
	"retained_earnings":             "Retained Earnings",
	"common_stock":                  "Common Stock",
}

// Common cash flow statement field mappings
var CashFlowFields = map[string]string{
	"net_cash_flow": "Net Change in Cash",
	"net_cash_flow_from_operating_activities": "Operating Cash Flow",
	"net_cash_flow_from_investing_activities": "Investing Cash Flow",
	"net_cash_flow_from_financing_activities": "Financing Cash Flow",
	"depreciation_and_amortization":           "Depreciation & Amortization",
	"capital_expenditure":                     "Capital Expenditure",
	"purchase_of_investment_securities":       "Investment Purchases",
	"sale_of_investment_securities":           "Investment Sales",
	"payment_of_dividends":                    "Dividends Paid",
	"repurchase_of_common_stock":              "Stock Buybacks",
	"issuance_of_common_stock":                "Stock Issuance",
	"issuance_of_debt":                        "Debt Issuance",
	"repayment_of_debt":                       "Debt Repayment",
}

// Financial ratio field mappings
var RatioFields = map[string]string{
	"return_on_equity":           "Return on Equity (ROE)",
	"return_on_assets":           "Return on Assets (ROA)",
	"return_on_invested_capital": "Return on Invested Capital",
	"gross_margin":               "Gross Margin",
	"operating_margin":           "Operating Margin",
	"net_profit_margin":          "Net Profit Margin",
	"current_ratio":              "Current Ratio",
	"quick_ratio":                "Quick Ratio",
	"debt_to_equity":             "Debt to Equity",
	"debt_to_assets":             "Debt to Assets",
	"interest_coverage":          "Interest Coverage",
	"asset_turnover":             "Asset Turnover",
	"inventory_turnover":         "Inventory Turnover",
	"receivables_turnover":       "Receivables Turnover",
	"price_to_earnings":          "P/E Ratio",
	"price_to_book":              "P/B Ratio",
	"price_to_sales":             "P/S Ratio",
	"enterprise_value_to_ebitda": "EV/EBITDA",
}
